packages/data/src/hooks/use-select/index.js

```diff
@@ -0,0 +1,53 @@
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState, useContext } from '@wordpress/element';
+
+/**
+ * External dependencies
+ */
+import { forEach } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { RegistryContext } from '../../components/registry-provider';
+
+/**
+ * A factory returning a enhanced selector as a custom hook.
+ *
+ * @param {function} subscribe  The subscribe function for the current registry.
+ * @param {function} selector   The selector being enhanced.
+ *
+ * @return {function} The enhanced selector
+ */
+const selectorHookFactory = ( subscribe, selector ) => ( ...args ) => {
+	const [ result, setResult ] = useState( null );
+	useEffect( () => {
+		const unsubscribe = subscribe( () => {
+			setResult( selector( ...args ) );
+		} );
+		return () => unsubscribe();
```

Comment:
one thing I'd like to explore here once #15436 is available, is if the selector has a resolver and the return is `undefined` then don't set the state.  This would prevent unnecessary renders while the selector is resolving.  It effectively would normalize `undefined` returns to null for all selectors with resolvers.  That's one issue I've seen in the wild with `withSelect` currently in that unless the mapping function handles `undefined` results, re-renders are more frequent.